### ISSUE TITLE
TSAN: data race ControlPlaneClient

### ISSUE THREAD

**Issue by ejona86** [grpc/grpc-java#10368]
I _think_ this was triggered in `XdsClientFederationTest.lrsClientsHaveDistinctLoadStatsManagers`.

[Logs](https://fusion2.corp.google.com/ci/tap/grpc_java_external.tsan/activity/547084865/invocations/39e26892-cdcb-407b-baea-cf8d3920c178/targets/%2F%2Fthird_party%2Fjava_src%2Fgrpc%2Fxds:src%2Ftest%2Fjava%2Fio%2Fgrpc%2Fxds%2FXdsClientFederationTest/tests). 

```
  Write of size 4 at 0x0000cc187fc8 by thread T4 (mutexes: write M0, write M1, write M2, write M3, write M4, write M5, write M6, write M7, write M8, write M9):
    #0 io.grpc.xds.ControlPlaneClient$AbstractAdsStream.cleanUp()V ControlPlaneClient.java:369 
    #1 io.grpc.xds.ControlPlaneClient$AbstractAdsStream.close(Ljava/lang/Exception;)V ControlPlaneClient.java:363 
    #2 io.grpc.xds.ControlPlaneClient$1.run()V ControlPlaneClient.java:137 
    #3 io.grpc.SynchronizationContext.drain()V SynchronizationContext.java:94 
    #4 io.grpc.SynchronizationContext.execute(Ljava/lang/Runnable;)V SynchronizationContext.java:126 
    #5 io.grpc.xds.XdsClientImpl.shutdown()V XdsClientImpl.java:213 
    #6 io.grpc.xds.SharedXdsClientPoolProvider$RefCountedXdsClientObjectPool.returnObject(Ljava/lang/Object;)Lio/grpc/xds/XdsClient; SharedXdsClientPoolProvider.java:146 
    #7 io.grpc.xds.SharedXdsClientPoolProvider$RefCountedXdsClientObjectPool.returnObject(Ljava/lang/Object;)Ljava/lang/Object; SharedXdsClientPoolProvider.java:103 
    #8 io.grpc.xds.XdsClientFederationTest.cleanUp()V XdsClientFederationTest.java:86 

  Previous read of size 4 at 0x0000cc187fc8 by thread T27:
    #0 io.grpc.xds.ControlPlaneClient.isReady()Z ControlPlaneClient.java:210 
    #1 io.grpc.xds.ControlPlaneClient.readyHandler()V ControlPlaneClient.java:218 
    #2 io.grpc.xds.ControlPlaneClient$AdsStreamV3$1$$Lambda$102.run()V ?? 
    #3 io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onReady()V ClientCalls.java:498 
    #4 io.grpc.PartialForwardingClientCallListener.onReady()V PartialForwardingClientCallListener.java:44 
    #5 io.grpc.ForwardingClientCallListener.onReady()V ForwardingClientCallListener.java:23 
    #6 io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onReady()V ForwardingClientCallListener.java:40 
    #7 io.grpc.PartialForwardingClientCallListener.onReady()V PartialForwardingClientCallListener.java:44 
    #8 io.grpc.ForwardingClientCallListener.onReady()V ForwardingClientCallListener.java:23 
    #9 io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onReady()V ForwardingClientCallListener.java:40 
    #10 io.grpc.internal.DelayedClientCall$DelayedListener.onReady()V DelayedClientCall.java:497 
    #11 io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamOnReady.runInternal()V ClientCallImpl.java:774 
    #12 io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamOnReady.runInContext()V ClientCallImpl.java:765 
    #13 io.grpc.internal.ContextRunnable.run()V ContextRunnable.java:37 
    #14 io.grpc.internal.SerializingExecutor.run()V SerializingExecutor.java:133 
    #15 java.util.concurrent.ThreadPoolExecutor.runWorker(Ljava/util/concurrent/ThreadPoolExecutor$Worker;)V ThreadPoolExecutor.java:1130 
    #16 java.util.concurrent.ThreadPoolExecutor$Worker.run()V ThreadPoolExecutor.java:630 
    #17 java.lang.Thread.run()V Thread.java:830 

  Location is heap block of size 88 at 0x0000cc187f80 allocated by thread T4:
    #0 io.grpc.xds.XdsClientImpl$ResourceSubscriber.<init>(Lio/grpc/xds/XdsClientImpl;Lio/grpc/xds/XdsResourceType;Ljava/lang/String;)V XdsClientImpl.java:523 
    #1 io.grpc.xds.XdsClientImpl$4.run()V XdsClientImpl.java:305 
    #2 io.grpc.SynchronizationContext.drain()V SynchronizationContext.java:94 
    #3 io.grpc.SynchronizationContext.execute(Ljava/lang/Runnable;)V SynchronizationContext.java:126 
    #4 io.grpc.xds.XdsClientImpl.watchXdsResource(Lio/grpc/xds/XdsResourceType;Ljava/lang/String;Lio/grpc/xds/XdsClient$ResourceWatcher;)V XdsClientImpl.java:293 
    #5 io.grpc.xds.XdsClientFederationTest.lrsClientsHaveDistinctLoadStatsManagers()V XdsClientFederationTest.java:195 
```

CC @temawi, @YifeiZhuang

**Comment by ejona86** [comment]
[Similar failure](https://fusion2.corp.google.com/invocations/7511a53c-934d-4c7b-8ced-a7eafa48ef45):

```
  Read of size 4 at 0x0000cff6ea48 by thread T22:
    #0 io.grpc.xds.ControlPlaneClient.isReady()Z ControlPlaneClient.java:210 
    #1 io.grpc.xds.ControlPlaneClient.readyHandler()V ControlPlaneClient.java:218 
    #2 io.grpc.xds.ControlPlaneClient$AdsStreamV3$1AdsClientResponseObserver$$Lambda$102.run()V ?? 
    #3 io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onReady()V ClientCalls.java:488 
    #4 io.grpc.PartialForwardingClientCallListener.onReady()V PartialForwardingClientCallListener.java:44 
    #5 io.grpc.ForwardingClientCallListener.onReady()V ForwardingClientCallListener.java:23 
    #6 io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onReady()V ForwardingClientCallListener.java:40 
    #7 io.grpc.PartialForwardingClientCallListener.onReady()V PartialForwardingClientCallListener.java:44 
    #8 io.grpc.ForwardingClientCallListener.onReady()V ForwardingClientCallListener.java:23 
    #9 io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onReady()V ForwardingClientCallListener.java:40 
    #10 io.grpc.internal.DelayedClientCall$DelayedListener.onReady()V DelayedClientCall.java:497 
    #11 io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamOnReady.runInternal()V ClientCallImpl.java:774 
    #12 io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamOnReady.runInContext()V ClientCallImpl.java:765 
    #13 io.grpc.internal.ContextRunnable.run()V ContextRunnable.java:37 
    #14 io.grpc.internal.SerializingExecutor.run()V SerializingExecutor.java:133 
    #15 java.util.concurrent.ThreadPoolExecutor.runWorker(Ljava/util/concurrent/ThreadPoolExecutor$Worker;)V ThreadPoolExecutor.java:1130 
    #16 java.util.concurrent.ThreadPoolExecutor$Worker.run()V ThreadPoolExecutor.java:630 
    #17 java.lang.Thread.run()V Thread.java:830 
    #18 (Generated Stub) <null> 

  Previous write of size 4 at 0x0000cff6ea48 by thread T4 (mutexes: write M0, write M1, write M2, write M3, write M4, write M5):
    #0 io.grpc.xds.ControlPlaneClient$AbstractAdsStream.cleanUp()V ControlPlaneClient.java:369 
    #1 io.grpc.xds.ControlPlaneClient$AbstractAdsStream.close(Ljava/lang/Exception;)V ControlPlaneClient.java:363 
    #2 io.grpc.xds.ControlPlaneClient$1.run()V ControlPlaneClient.java:137 
    #3 io.grpc.SynchronizationContext.drain()V SynchronizationContext.java:94 
    #4 io.grpc.SynchronizationContext.execute(Ljava/lang/Runnable;)V SynchronizationContext.java:126 
    #5 io.grpc.xds.XdsClientImpl.shutdown()V XdsClientImpl.java:213 
    #6 io.grpc.xds.SharedXdsClientPoolProvider$RefCountedXdsClientObjectPool.returnObject(Ljava/lang/Object;)Lio/grpc/xds/XdsClient; SharedXdsClientPoolProvider.java:146 
    #7 io.grpc.xds.SharedXdsClientPoolProvider$RefCountedXdsClientObjectPool.returnObject(Ljava/lang/Object;)Ljava/lang/Object; SharedXdsClientPoolProvider.java:103 
    #8 io.grpc.xds.XdsClientFederationTest.cleanUp()V XdsClientFederationTest.java:86 
```

**Comment by ejona86** [comment]
This was fixed by #10945
